plenum/server/future_primaries.py

```diff
@@ -0,0 +1,82 @@
+from typing import Dict, List
+
+from common.exceptions import LogicError
+from plenum.common.constants import TXN_TYPE, NODE, TARGET_NYM, ALIAS, SERVICES, VALIDATOR
+from plenum.common.util import getMaxFailures
+from plenum.server.batch_handlers.three_pc_batch import ThreePcBatch
+
+
+class FuturePrimaries:
+    # This class is needed for correct primaries storing in audit ledger.
+    # It is something like uncommitted state of primaries.
+    # It emulate primaries changes when pool txn is just applied.
+    # When current batches reverts, it drops it's properties to node's.
+    # TODO: when pluggable request handler implemented, make this class in the same style
+
+    def __init__(self, node, node_names, ordered_node_ids, current_required):
```

Comment:
In order to avoid duplicate code, I would make this part of PrimarySelector class as follows:
- Primary selector should have a class `calculate_primaries(nodeReg)` which can calculate (but not set!) next primaries
- `select_primaries` in Node can just call this method for the current nodeReg and apply (set) the new primaries for each replica
- AuditBatchHanler can have a reference to PrimarySelector.
- When preparing audit txn, it can do the following:
  - for every valid txn check that this is a Node txn leading to changes of F
  - If so, cope the current nodeReg and add/remove a new node from it
  - pass nodeReg to `calculate_primaries` and get the future primaries

Reply:
Implemented